### PR TITLE
Fix hang in BackoffDiscovery.FindPeers when requesting limit lower than number of peers available

### LINKED ISSUE #67: Possible deadlock in FindPeers
Hi, I don't know if this library still maintained, but if `chLen` is smaller than `c.prevPeers` a deadlock can occur [here](https://github.com/libp2p/go-libp2p-discovery/blob/master/backoffcache.go#L142) due to a full buffered channel,  no?

I suppose that the comparison in this [line](https://github.com/libp2p/go-libp2p-discovery/blob/master/backoffcache.go#L137)  was intended to be the inverse logic: `chLen < len(c.prevPeers)`, which fix the deadlock.

### PR DESCRIPTION
**Pull request by iand**
Fixes #67

### CODE REVIEW

File: backoffcache.go [hunk 1]
```diff
@@ -139,7 +140,11 @@ func (d *BackoffDiscovery) FindPeers(ctx context.Context, ns string, opts ...dis
 		}
 		pch := make(chan peer.AddrInfo, chLen)
 		for _, ai := range c.prevPeers {
-			pch <- ai
+			select {
+			case pch <- ai:
+			default:
+				// skip if we have asked for a lower limit than the number of peers known
```
Comment (marten-seemann) [inline backoffcache.go:146]: I'm wondering if it's ok here to return more than `options.Limit` values. The consumer could rely on this limit, as it's documented here:
https://github.com/libp2p/go-libp2p-core/blob/525a0b13017263bde889a3295fa2e4212d7af8c5/discovery/options.go#L35-L41

**Comment by BigLep** [comment]
@iand : can you please incorporate comments so we can get this merged?

File: backoffcache.go (hunk 1 above)
Comment (iand) [reply backoffcache.go:146]: Can you clarify your comment, it seems ambiguous and I am not sure if you are asking if it's ok to do something different to this code or if this code is ok as is. This code respects the limit by skipping any that exceed the number requested.

Comment (marten-seemann) [reply backoffcache.go:146]: Sorry, I don't understand any more what I was saying 4 weeks ago either ;)

**Review by marten-seemann — APPROVED** [review]
LGTM.

**Comment by BigLep** [comment]
@marten-seemann : can you please merge this and get it bubbled to go-ipfs?

**Pull request merged**